repo/migrations/Migration012.go

```diff
@@ -0,0 +1,261 @@
+	if err != nil {
+		return err
+	}
+
+	listingsIndex := []*Migration012_ListingData{}
+	err = json.Unmarshal(listingsIndexJSONBytes, &listingsIndex)
+	if err != nil {
+		return err
+	}
+
+	cryptoListings := []*Migration012_ListingData{}
+	for _, listingAbstract := range listingsIndex {
+		if listingAbstract.ContractType == "CRYPTOCURRENCY" {
+			cryptoListings = append(cryptoListings, listingAbstract)
+		}
+	}
+
+	// Finish early If no crypto listings
+	if len(cryptoListings) == 0 {
+		return writeRepoVer(repoPath, 13)
+	}
+
+	// Check each crypto listing for markup
+	markupListings := []*pb.SignedListing{}
+	for _, listingAbstract := range cryptoListings {
+		listingJSONBytes, err := ioutil.ReadFile(migration012_listingFilePath(repoPath, listingAbstract.Slug))
+		if err != nil {
+			return err
+		}
+		sl := new(pb.SignedListing)
+		err = jsonpb.UnmarshalString(string(listingJSONBytes), sl)
+		if err != nil {
+			return err
+		}
+
+		if Migration012_listingHasNewFeaturesAndOldVersion(sl) {
+			markupListings = append(markupListings, sl)
+		}
+	}
+
+	// Finish early If no crypto listings with new features are found
+	if len(markupListings) == 0 {
+		return writeRepoVer(repoPath, 13)
+	}
+
+	// Setup signing capabilities
+	identityKey, err := Migration012_GetIdentityKey(repoPath, databasePassword, testnetEnabled)
+	if err != nil {
+		return err
+	}
+
+	identity, err := ipfs.IdentityFromKey(identityKey)
+	if err != nil {
+		return err
+	}
+
+	// IPFS node setup
+	r, err := fsrepo.Open(repoPath)
+	if err != nil {
+		return err
+	}
+	cctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	cfg, err := r.Config()
+	if err != nil {
+		return err
+	}
+
+	cfg.Identity = identity
+
+	ncfg := &ipfscore.BuildCfg{
+		Repo:   r,
+		Online: true,
+		ExtraOpts: map[string]bool{
+			"mplex": true,
+		},
+		DNSResolver: nil,
+		Routing:     nil,
+	}
+
+	nd, err := ipfscore.NewNode(cctx, ncfg)
+	if err != nil {
+		return err
+	}
+	defer nd.Close()
+
+	// Update each listing to have the latest version number and resave
+	// Save the new hashes for each changed listing so we can update the index.
+	hashes := make(map[string]string)
+
+	for _, sl := range markupListings {
+		sl.Listing.Metadata.Version = migration012_ListingVersionForMarkupListings
+
+		serializedListing, err := proto.Marshal(sl.Listing)
+		if err != nil {
+			return err
+		}
+		idSig, err := nd.PrivateKey.Sign(serializedListing)
+		if err != nil {
+			return err
+		}
+		sl.Signature = idSig
+
+		m := jsonpb.Marshaler{
+			EnumsAsInts:  false,
+			EmitDefaults: false,
+			Indent:       "    ",
+			OrigName:     false,
+		}
+		out, err := m.MarshalToString(sl)
+		if err != nil {
+			return err
+		}
+
+		filename := migration012_listingFilePath(repoPath, sl.Listing.Slug)
+		if err := ioutil.WriteFile(filename, []byte(out), os.ModePerm); err != nil {
+			return err
+		}
+		h, err := ipfs.GetHashOfFile(nd, filename)
```

Comment:
Is it worth ensuring this function produces a specific result in a test? If the lib changes the hashing, that test could alert us to the problem migration.